### PR TITLE
Create `UserCredentials` entity

### LINKED ISSUE #209: Create `UserCredentials` entity
## Description
Create `UserCredentials` entity for storing user credentials safely

### Requirements
- project should be runnable

### Deliverables
- `UserCredentials` entity, must be owned type of `User`
- `UserCredentialsConfiguration` entity configuration
- `UserCredentailsRepository` repository
- `UserCredentialsService` foundations service
- fluent validation
- add migration
- register all necessary infrastructure to DI

### PR DESCRIPTION
**Pull request by Mirzohid-Mominov**
closes #209

### CODE REVIEW

File: src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/DataContexts/AppDbContext.cs [hunk 1]
```diff
@@ -17,6 +17,8 @@ public class AppDbContext(DbContextOptions<AppDbContext> options) : DbContext(op
 
     public DbSet<UserSettings> UserSettings => Set<UserSettings>();
 
+    public DbSet<UserCredentials> UserCredentialss => Set<UserCredentials>();
+
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/DataContexts/AppDbContext.cs:21]: Remove it, as `UserCredentials` isn't a separate entity

File: src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Migrations/20240130111752_Add_UserCredentials.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                {
+                    table.PrimaryKey("PK_UserCredentialss", x => x.UserId);
+                    table.ForeignKey(
+                        name: "FK_UserCredentialss_Users_UserId",
+                        column: x => x.UserId,
+                        principalTable: "Users",
+                        principalColumn: "Id",
+                        onDelete: ReferentialAction.Cascade);
+                });
+        }
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Migrations/20240130111752_Add_UserCredentials.cs:35]: `UserCredentials` is added as separate table here, need to re-generate migrations

File: src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/UserValidator.cs [hunk 3]
```diff
@@ -45,12 +45,6 @@ public UserValidator(IOptions<ValidationSettings> validationSettings)
                     .MaximumLength(64)
                     .Matches(validationSettingsValue.NameRegexPattern)
                     .WithMessage("Last name is not valid");
-
-                RuleFor(user => user.PasswordHash)
-                    .NotEmpty()
-                    .MinimumLength(8)
-                    .MaximumLength(64)
-                    .Matches(validationSettingsValue.PasswordRegexPattern);
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/UserValidator.cs:53]: Move this to `UserCredentialsValidator`

**Review by sulton-max — CHANGES_REQUESTED** [review]

File: src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/UserValidator.cs (hunk 3 above)
Comment (Mirzohid-Mominov) [reply src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/UserValidator.cs:53]: done

**Review by Abdurasulovich — COMMENTED** [review]
Done

File: src/Training.AirBnb.Clone.Backend/AirBnB.Domain/Entities/User.cs [hunk 4]
```diff
@@ -61,4 +56,6 @@ public sealed class User : AuditableEntity
     /// Gets or sets listings owned by the user.
     /// </summary>
     public List<Listing> Listings { get; set; }
+
```
Comment (rashidovasadbek) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Domain/Entities/User.cs:59]: you need to write a comment

File: src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/UserCredentialsValidator.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        
+        RuleSet(
+            EntityEvent.OnCreate.ToString(),
+            () =>
+            {
+                RuleFor(userCredentials => userCredentials.PasswordHash)
+                    .NotEmpty()
+                    .MinimumLength(8)
+                    .MaximumLength(64)
+                    .Matches(validationSettingsValue.PasswordRegexPattern);
```
Comment (Ilxom-22) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/UserCredentialsValidator.cs:24]: Do we really need to validate hashed password?

**Review by Ilxom-22 — CHANGES_REQUESTED** [review]

File: src/Training.AirBnb.Clone.Backend/AirBnB.Domain/Entities/User.cs (hunk 4 above)
Comment (Mirzohid-Mominov) [reply src/Training.AirBnb.Clone.Backend/AirBnB.Domain/Entities/User.cs:59]: I think it's not mine

File: src/Training.AirBnb.Clone.Backend/AirBnB.Api/Data/SeedDataExtensions.cs [hunk 6]
```diff
@@ -101,7 +101,10 @@ private static async ValueTask SeedUsersAsync(this AppDbContext dbContext)
             .RuleFor(user => user.FirstName, data => data.Name.FirstName())
             .RuleFor(user => user.LastName, data => data.Name.LastName())
             .RuleFor(user => user.EmailAddress, data => data.Person.Email)
-            .RuleFor(user => user.PasswordHash, data => data.Internet.Password(8))
+            .RuleFor(user => user.UserCredentials, data => new UserCredentials
+            {
+                PasswordHash = data.Internet.Password(8)
```
[the remark this answers is not in the document]
Comment (Mirzohid-Mominov) [reply src/Training.AirBnb.Clone.Backend/AirBnB.Api/Data/SeedDataExtensions.cs:106]: got it

File: src/Training.AirBnb.Clone.Backend/AirBnB.Api/Data/SeedDataExtensions.cs [hunk 7]
```diff
@@ -114,7 +117,10 @@ private static async ValueTask SeedUsersAsync(this AppDbContext dbContext)
             .RuleFor(user => user.FirstName, data => data.Name.FirstName())
             .RuleFor(user => user.LastName, data => data.Name.LastName())
             .RuleFor(user => user.EmailAddress, data => data.Person.Email)
-            .RuleFor(user => user.PasswordHash, data => data.Internet.Password(8))
+            .RuleFor(user => user.UserCredentials, data => new UserCredentials()
+            {
+                PasswordHash = data.Internet.Password(8)
```
[the remark this answers is not in the document]
Comment (Mirzohid-Mominov) [reply src/Training.AirBnb.Clone.Backend/AirBnB.Api/Data/SeedDataExtensions.cs:122]: it's hashed we did with teacher

**Review by sulton-max — APPROVED** [review]
Looks good to me

**Review by Abdurasulovich — COMMENTED** [review]
Done

**Review by Ilxom-22 — APPROVED** [review]

**Pull request merged**